resourcemanager/scheduler/gradient2_scheduler.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scheduler
+
+import (
+	"time"
+
+	"github.com/pingcap/tidb/resourcemanager/util"
+	"github.com/pingcap/tidb/util/mathutil"
+)
+
+const (
+	defaultSmoothing      float64 = 0.2
+	defaultMinConcurrency int     = 2
+
+	maxOverloadConcurrencyDelta int = 2 // max concurrency = user setting concurrency + delta
+)
+
+// Gradient2Scheduler is a scheduler using Gradient2 algorithm.
+// Gradient2 attempts to reduce the delay by obeserving the long and short window RTT and queue length to predict
+// reasonable concurrency. Using average algorithm to smooth the result for making scheduler more robust.
+type Gradient2Scheduler struct {
+	smoothing float64
+}
+
+// NewGradient2Scheduler is to create a new Gradient2Scheduler
+func NewGradient2Scheduler() *Gradient2Scheduler {
+	return &Gradient2Scheduler{
+		smoothing: defaultSmoothing,
+	}
+}
+
+// Tune is to tune the concurrency of the component
+func (b *Gradient2Scheduler) Tune(c util.Component, p util.GorotinuePool) Command {
+	newLimit := b.tune(c, p)
+	if newLimit > float64(p.Running()) {
+		return Overclock
+	} else if newLimit < float64(p.Running()) {
+		return Downclock
+	}
+	return Hold
+}
+
+func (b *Gradient2Scheduler) tune(_ util.Component, p util.GorotinuePool) float64 {
+	if time.Since(p.LastTunerTs()) < minCPUSchedulerInterval {
```

Comment:
As our discussion， Please add one system variable to let user could decide the tuning frequence, since it will affect performance according the setting. also we should give a default value that will affect performance less than 5% and describe in this value how long will system react the workload change.

Reply:
Yes, We can add the system variable at the end. It is not suitable for the developing function. I can add TODO.

Reply:
```minCPUSchedulerInterval``` has been changed into atomic type so that we can dynamically change it.